src/server/active_tran_server.hpp

```diff
@@ -45,7 +47,11 @@ class active_tran_server : public tran_server
   private:
     bool m_uses_remote_storage = false;
 
-    cublog::prior_sender::sink_hook_t m_prior_sender_sink_hook_func;
+    /*
+     * sends prior nodes to the page servers.
+     * The order can differ from m_page_server_conn_vec
+     */
+    std::list<cublog::prior_sender::sink_hook_t> m_prior_sender_sink_hooks;
```

Comment:
@cristiarg I don't use `std::vector` because the functor `sink_hook_t` has no `noexcept` move constructor so the vector copies them when it grows (until c++20), which in turn `prior_sender:: m_sink_hooks` may refer to a removed object.

Please share your opinion if you have a better idea.

Reply:
However, this is more complicated here.

I think we need a design of objects that is able to fulfill the following functional requirements:
- a PTS must be able to connect to multiple PS's, disconnect from some (when those Ps's become unavailable; connect to new ones when those are "advertised" to the cluster; ... and whatnot ..)
- a PTS must be able to fail-over as an ATS
- but, at the same time, a PTS will not need to send prior log data to the connected PS's (this is subjective, the functionality "could" be formally hidden from the PTS but does not need to be - we'll not do this as it over-complicates things even further)
- an ATS must be able to send prior log data to connected PS's
- an ATS (and maybe PTS's too) must be able to interrogate the connected PS's about all kinds of information that will be needed to implement the consensus logic
- ...

We have more than one way to achieve the above:

1. Simple way:
We need a way to identify the same PS across different structures (ie: not by index as it is now)
We have the following options:
- by `cubcomm::channe::get_channel_id` (as is done with `page_server::connection_handler::get_connection_id`)
- by `cubcomm::node` information - "host:port" (although only host might be enough)
Then we can use this id to juggle different maps

2. More formal way:
I think we would better refactor to group connection information to PS's in some sort of `connection_handler` as is done with `page_server::connection_handler`.

@wrlawodms  
This patch is only preliminary and could be merged - up to you.
But we also nee to refactor for future development either with one of the options above or with one proposal of yours.
Let me know your opinion. Thanks.

Reply:
> More formal way:
I think we would better refactor to group connection information to PS's in some sort of connection_handler as is done with page_server::connection_handler.

1. Actually, this is the way I mentioned in the previous meeting. I am going to deal with disconnection cases in the following issue: http://jira.cubrid.org/browse/LETS-639. How about refactoring by adding connection_handler in the issue? 
2. And I'm planning to only add the connection handler for ATS, not PTS. I doubt that the PTS also requires this kind of class, but we can add later if needed.

Please share your opinion on it.

> maybe you can use unique_ptr as is done in tran_server::m_page_server_conn_vec

I considered it, but unique_ptr wouldn't work for this case, because the prior sender also points to sinks. We can use shared_ptr, but it requires to make changes to the prior_sender too, the way I see it. Anyway, If we go with the above "connection_handler" refactoring, we don't need to worry about it. Then, I'l like to go with `list` for now, and we can change it with the refactoring.

Reply:
> How about refactoring by adding connection_handler in the issue?

As I said, it can be done separately. This patch is self-contained as it is.

> And I'm planning to only add the connection handler for ATS, not PTS. I doubt that the PTS also requires this kind of class, but we can add later if needed.

Actually you are right.
I was concerned about the PTS fail-over to ATS in the future.
But for the fail-ove, a PTS will have to swap the `passive_tran_server` with the `active_tran_server` structure as a whole.

Reply:
> I considered it, but unique_ptr wouldn't work for this case, because the prior sender also points to sinks. We can use shared_ptr, but it requires to make changes to the prior_sender too, the way I see it. Anyway, If we go with the above "connection_handler" refactoring, we don't need to worry about it. Then, I'd like to go with list for now, and we can change it with the refactoring.

`unique_ptr` does work :)
The `active_tran_server` instance is the owner. `prior_sender::add_sink` takes as argument a reference (just for the formalism of ensuring that there is a valid instance) which is then referentiated as a pointer. And `prior_sender::add_sink` has non-owning pointers.

Reply:
> unique_ptr does work :)
Aha, OK. The prior_sender doesn't need to own this. I got it wrong. Thank you.